psl-core/src/main/java/org/linqs/psl/model/predicate/Predicate.java

```diff
@@ -45,6 +47,22 @@
 	 */
 	Predicate(String name, ConstantType[] types) {
 		this.types = types;
+		this.names = null;
+		predicateName = name.toUpperCase();
+	}
+
+	/**
+	 * Constructor.
+	 * <p>
+	 * Should only be called by {@link PredicateFactory}.
+	 *
+	 * @param name  name for this predicate
+	 * @param types  types for each of the predicate's arguments
+	 * @param names  names for each of the predicate's arguments
+	 */
+	Predicate(String name, ConstantType[] types, List<String> names) {
+		this.types = types;
+		this.names = names;
```

Comment:
assert that there are the correct number of names and make the list an immutable deep copy.
```
this.name = Collections.unmodifiableList(new ArrayList<String>(names));
```